pennylane/ops/op_math/adjoint.py

```diff
@@ -162,25 +166,84 @@ def loop_fn(i):
         available_eps = compiler.AvailableCompilers.names_entrypoints
         ops_loader = available_eps[active_jit]["ops"].load()
         return ops_loader.adjoint(fn, lazy=lazy)
-    if qml.math.is_abstract(fn):
-        return Adjoint(fn)
     return create_adjoint_op(fn, lazy)
 
 
 def create_adjoint_op(fn, lazy):
     """Main logic for qml.adjoint, but allows bypassing the compiler dispatch if needed."""
+    if qml.math.is_abstract(fn):
+        return Adjoint(fn)
     if isinstance(fn, Operator):
         return Adjoint(fn) if lazy else _single_op_eager(fn, update_queue=True)
-    if not callable(fn):
-        raise ValueError(
-            f"The object {fn} of type {type(fn)} is not callable. "
-            "This error might occur if you apply adjoint to a list "
-            "of operations instead of a function or template."
+    if callable(fn):
+        if qml.capture.enabled():
+            return _capture_adjoint_transform(fn, lazy=lazy)
+        return _adjoint_transform(fn, lazy=lazy)
+    raise ValueError(
+        f"The object {fn} of type {type(fn)} is not callable. "
+        "This error might occur if you apply adjoint to a list "
+        "of operations instead of a function or template."
+    )
+
+
+@lru_cache  # only create the first time requested
+def _get_adjoint_qfunc_prim():
+    """See capture/explanations.md : Higher Order primitives for more information on this code."""
+    # if capture is enabled, jax should be installed
+    import jax  # pylint: disable=import-outside-toplevel
+
+    AbstractOperator = qml.capture.AbstractOperator
+
+    adjoint_prim = jax.core.Primitive("adjoint_transform")
+    adjoint_prim.multiple_results = True
+
+    @adjoint_prim.def_impl
+    def _(*args, jaxpr, lazy=True):
+        with qml.queuing.AnnotatedQueue() as q:
+            jax.core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, *args)
+        ops, _ = qml.queuing.process_queue(q)
+        return [adjoint(op, lazy=lazy) for op in ops]
+
+    def _is_queued_outvar(outvars):
+        if not outvars:
+            return False
+        return isinstance(outvars[0].aval, AbstractOperator) and isinstance(
+            outvars[0], jax.core.DropVar
         )
 
-    @wraps(fn)
+    @adjoint_prim.def_abstract_eval
+    def _(*_, jaxpr, **__):
+        # all queued drop var operators
+        outvars = [AbstractOperator() for eqn in jaxpr.eqns if _is_queued_outvar(eqn.outvars)]
```

Comment:
Do we want to continue having the qfunc transform return all the operations? This might not scale well to bigger problems.

We may also end up with transforms that don't actually return `AbstractOperator` objects for every operator that is queued.  For example, `for_loop`.  In that situation, the concrete implementation will get queued operators that don't show up as `AbstractOperator`'s in the jaxpr.

Question:

Should we have `adjoint(qfunc)(*args, **kwargs)` just return an empy list?

Reply:
> Do we want to continue having the qfunc transform return all the operations? This might not scale well to bigger problems.

I do not think that's a good idea. If we want we can have adjoint return one operator (itself), or return nothing. Considering it now has an attached scoped which can be rather arbitrary (including additional adjoints or even loops), I would consider the adjoint version of each base operator to not exist yet.

> We may also end up with transforms that don't actually return AbstractOperator objects for every operator that is queued. For example, for_loop. In that situation, the concrete implementation will get queued operators that don't show up as AbstractOperator's in the jaxpr.

Yes this is another issue, since some "nesting primitives" need to be able return classical values (like for and cond), those cannot really return abstract operator instances (or maybe they could be returned alongside the classical values?).

With this in mind, I'm still uncertain about wether it would be better for adjoint/control to return just themselves, or nothing.

Either way, this would also be a good opportunity to unify the behaviour for how adjoint and control behave in PennyLane, since they currently return different things.

Reply:
`()` or `[]`, since those are essentially returning nothing. As far as I know, there is no "Symbolic None", and the abstract evaluation rule does not accept `None`.

Reply:
Sorry maybe I should have phrased this better. What I mean by symbolic adjoint is an operator instance, in abstract evaluation it would be of `AbstractOperator` type, in concrete evaluation it would be of `Adjoint` type. By symbolic I meant that it is not a "concrete" operator yet. It is an operator that represents one or more operators, but which operators exactly it represents is not defined yet. The reason it's not defined yet is because it is defined by a region of code which hasn't been evaluated yet (the body). 

Concretely:
```py
def qfunc():
  qml.RY(0.1, wires=0)
  qml.RX(0.2, wires=0)
  
>>> qml.adjoint(qfunc)()
Adjoint
```

Either way, we landed on returning nothing which is good, just wanted to clarify :)